aptos-move/mvhashmap/src/versioned_aggregators.rs

```diff
@@ -376,6 +376,42 @@ impl<K: Eq + Hash + Clone + Debug + Copy> VersionedAggregators<K> {
         );
     }
 
+    /// Must be called when an snapshot (delta or derived) creation with a given ID and initial apply is observed
+    /// This should be only called when apply applies on top of different ID.
+    pub fn create_dependent_aggregator(
+        &self,
+        id: K,
+        txn_idx: TxnIndex,
+        apply: AggregatorApplyChange<K>,
+    ) {
+        let mut created = VersionedValue::new(None);
+        created.insert(txn_idx, VersionEntry::Apply(apply));
+
+        assert_none!(
+            self.values.insert(id, created),
+            "VerionedValue when creating aggregator ID may not already exist"
+        );
+    }
+
+    pub fn record_change(&self, id: K, txn_idx: TxnIndex, change: AggregatorChange<K>) {
+        match change {
+            AggregatorChange::Create(value) => self.create_aggregator(id, txn_idx, value),
+            AggregatorChange::Apply(apply) => match &apply {
+                AggregatorApplyChange::AggregatorDelta { .. } => {
+                    self.values
+                        .get_mut(&id)
+                        // TODO we probably cannot panic here any more (i.e. in V2 this might not be guaranteed)
+                        .expect("VersionedValue for an (resolved) ID must already exist")
```

Comment:
Hm, why do we have this invariant?

Reply:
this was in the code below already, added comment that we should probably remove it.

We theoretically cannot have an Apply before a Create - as Create defines a new identifier. 
But because there is no order in how things are inserted into MVHashMap, future transaction can see a partial view from the previous transaction - can get an id from a resource write it sees, and produce an apply, before the Create is inserted into MVHashMap

Reply:
Oh, I remember now.
Even if we remove the create, we don't delete the VersionedValue (i.e. the versioned data-structure corresponding to the ID), and there should be no way for anyone to see a resource before it is added to the MVHashMap. As long as we record aggregator updates before writes, we could maintain this invariant.

When a transaction aborts, and we remove its create, the aggregator read will not be resolved, and this we can treat as a speculative error (replaced from storage or not, we expect things to get resolved)?

Reply:
"and there should be no way for anyone to see a resource before it is added to the MVHashMap." 

this is not true any more - "resource that contains aggregator ID" can be inserted before "create of aggregator" - so that needs to be a speculative NotFound error.

Reply:
isn't it up to us if we wanted to fix the order of MVHashMap insertions? not saying we do it, just for discussion.